### PR TITLE
Viite 673 + 470 + 682 termination and new change to beginning of road part transfer to the rest

### PR DESCRIPTION
**Pull request by plourenco90**
Contains tests for the 673
I merged with the 470 due to test errors that where fixed in the 470 branch

### CODE REVIEW

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/ProjectDeltaCalculator.scala [hunk 1]
```diff
@@ -83,8 +95,24 @@ object ProjectDeltaCalculator {
...
+    ).toSeq
+
+    (addressesGroups, projectLinksGroups)
+  }
+
 }
 
 case class Delta(startDate: DateTime, terminations: Seq[RoadAddress], newRoads: Seq[ProjectLink],
-                 unChanged: Seq[RoadAddress])
+                 unChanged: Seq[RoadAddress], transferredOld: Seq[RoadAddress], transferredNew: Seq[ProjectLink])
```
Comment (tvenhola) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/ProjectDeltaCalculator.scala]: Maybe transferred: (Seq[RoadAddress], Seq[ProjectLink])?

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/ProjectDeltaCalculator.scala [hunk 2]
```diff
@@ -83,8 +95,24 @@ object ProjectDeltaCalculator {
     ).toSeq
   }
 
+  def partition(roadAddresses: Seq[RoadAddress], projectLinks : Seq[ProjectLink]): (Seq[RoadAddressSection],Seq[RoadAddressSection]) = {
```
Comment (tvenhola) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/ProjectDeltaCalculator.scala]: Scaladoc for intended functionality - "Calculates the transfer sections on a road part and maps them together for change table"

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/ProjectDeltaCalculator.scala [hunk 3]
```diff
@@ -83,8 +95,24 @@ object ProjectDeltaCalculator {
...
+    val groupedAddresses = roadAddresses.sortBy(_.startAddrMValue).groupBy(ra => (ra.roadNumber, ra.roadPartNumber, ra.track))
+    val addressesGroups = groupedAddresses.mapValues(v => combine(v.sortBy(_.startAddrMValue))).values.flatten.map(ra =>
```
Comment (tvenhola) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/ProjectDeltaCalculator.scala]: Will fail if there are multiple, possibly gapped transfer sections on road.

Instead, map by link id and group then.
RoadAddress.linkId == ProjectLink.linkId on transfer.

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala [hunk 4]
```diff
@@ -708,18 +699,23 @@ class ProjectService(roadAddressService: RoadAddressService, roadLinkService: Ro
...
-      if (linkStatus == LinkStatus.Terminated)
-        ProjectDAO.updateProjectLinksToDB(updatedProjectLinks.map(_.copy(status=linkStatus, calibrationPoints = (None, None),
-          startAddrMValue = 0L, endAddrMValue = 0L)), userName)
+      val (updatedProjectLinks, unchangedProjectLinks) = projectLinks.filterNot(pl=> pl.status == LinkStatus.Terminated ).partition(pl => linkIds.contains(pl.linkId))
+      if (linkStatus == LinkStatus.Terminated) {
+        //Fetching road addresses in order to obtain the original addressMValues, since we may not have those values on project_link table, after previous recalculations
+        val roadAddresses = RoadAddressDAO.fetchByLinkId(updatedProjectLinks.map(pl => pl.linkId).toSet)
+        val updatedPL = updatedProjectLinks.map(pl => {
+          val roadAddress = roadAddresses.find(_.linkId == pl.linkId)
+          pl.copy(startAddrMValue = roadAddress.get.startAddrMValue, endAddrMValue = roadAddress.get.endAddrMValue)
```
Comment (Campanha) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala]: roadAddress could be empty here in anomalous cases here "val roadAddress = roadAddresses.find(_.linkId == pl.linkId)" is None. The "roadAddress.get" could give problems

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/ProjectDeltaCalculator.scala (hunk 3 above)
Comment (tvenhola) [reply digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/ProjectDeltaCalculator.scala]: Could also return a Map[RoadAddressSection, RoadAddressSection] which would give the correct mapping row immediately

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala (hunk 4 above)
Comment (tvenhola) [reply digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala]: Termination and transfer has always a link with a link id that remains the same. Should there not be one then we have problems elsewhere in the code.

Also VIITE-697 will affect this and make sure that one road address row is found for a project link (unless LinkStatus == New)

**Pull request merged**